### PR TITLE
Implement peer selection algorithm for block sync mechanism - Closes #3767

### LINKED ISSUE #3767: Implement peer selection algorithm step for Block Sync Mechanism
### Expected behavior
Implement peer selection algorithm for Block Sync Mechanism as stated in [LIP-0014](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#block-synchronization-mechanism)

Requires: #3774 

Request a list of all connected peers from the P2P module including T.height and T.heightPrevoted of the block T at the tip of their chain. From this information, determine a set of bestPeers that are at the tip of the chain according to the fork choice rule as follows:

a) Among all peers, choose the subset of peers with largest `T.heightPrevoted`.

b) From the peers in a) choose those with largest `T.height`.

c) Form the peers in b) choose the largest set which has the same `blockID`. Ties are broken in favor of smaller `blockID`. This set is the set `bestPeers` and we define `B` to be the block at the tip of the chain of this set of peers.

If the following expression evaluates to `false`, the block synchronization mechanism is aborted:

    (A.heightPrevoted<B.heightPrevoted) || (A.height<B.height && A.heightPrevoted==B.heightPrevoted)

### Which version(s) does this affect? (Environment, OS, etc...)
`feature/BFT`

### PR DESCRIPTION
**Pull request by limiaspasdaniel**
### What was the problem?
Peer selection algorithm for Block Synchronization Mechanism was not implemented
### How did I fix it?
By implementing the peer selection algorithm defined in LIP-004 step 1 of [Block Synchronization Mechanism](https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#block-synchronization-mechanism)
### Review checklist

* The PR resolves #3767
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

File: framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js [hunk 1]
```diff
@@ -1,14 +1,114 @@
...
+	 * @example
+	 *
+	 * Input: [{id: 1, height: 2}, {id: 2, height: 3}, {id: 3, height: 3}]
+	 * Output: [{id: 2, height: 3}, {id: 3, height: 3}]
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_computeLargestSubsetMaxBy(arrayOfObjects, condition) {
+		const absoluteMax = condition(maxBy(arrayOfObjects, condition));
+		const largestSubset = [];
+		// eslint-disable-next-line no-restricted-syntax
```
Comment (yatki) [inline framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js]: Not part of the PR but we can disable this rule or modify it to ignore `BinaryExpression[operator='of']`.

Comment (limiaspasdaniel) [reply framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js]: As discussed in the meeting we have to come up with a conclusion on this topic :) Resolving for now

File: framework/src/controller/application_state.js [hunk 2]
```diff
@@ -85,16 +85,18 @@ class ApplicationState {
...
 	 */
-	async update({ height, prevotedConfirmedUptoHeight, broadhash }) {
+	async update({ height, prevotedConfirmedUptoHeight, broadhash, blockId }) {
 		assert(height, 'height is required to update application state.');
 		try {
 			const newState = this.state;
 			newState.broadhash = broadhash || this.state.broadhash;
 			newState.prevotedConfirmedUptoHeight =
 				prevotedConfirmedUptoHeight || this.state.prevotedConfirmedUptoHeight;
+			newState.blockId = blockId || this.state.blockId;
```
[the remark this answers is not in the document]
Comment (limiaspasdaniel) [reply framework/src/controller/application_state.js]: > maybe we can set default values for the params like?:
> 
> ```js
> async update({ height, prevotedConfirmedUptoHeight, broadhash, blockId = this.state.blockId })
> ```
> 
> same goes for other params.

Agreed, definitely cleaner, thanks.

[the remark this answers is not in the document]
Comment (limiaspasdaniel) [reply framework/src/controller/application_state.js]: > Also, I know it's out of this PR's scope, not relevant what you did but the whole update function can be simplified as below ¯\_(ツ)_/¯ :
> 
> ```js
> async update(params) {
>  const newState = {...this.state, ...params};
> __private.state.set(this, newState);
> this.logger.debug('Application state', this.state);	
> await this.stateChannel.publish('app:state:updated', this.state);
> return true;
> }
> ```
> 
> Also, I don't like catching errors just to log them.

Agreed, for that, I created an issue to improve it: #3875

File: elements/lisk-p2p/src/p2p_types.ts [hunk 3]
```diff
@@ -62,6 +62,8 @@ export interface P2PNodeInfo {
 	readonly protocolVersion: string;
 	readonly nethash: string;
 	readonly wsPort: number;
+	readonly blockId: string;
```
Comment (nazarhussain) [inline elements/lisk-p2p/src/p2p_types.ts]: Its too generic, we must make it sepcific for understanding. `lastBlockId`, `finalizedBlockId` or any thing which is specifically been communicated here.

Comment (limiaspasdaniel) [reply elements/lisk-p2p/src/p2p_types.ts]: Agreed, will change it to `lastBlockId`

Comment (limiaspasdaniel) [reply elements/lisk-p2p/src/p2p_types.ts]: Done

File: framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js [hunk 4]
```diff
@@ -1,14 +1,116 @@
...
+	 * @return {Array<Object>}
+	 * @private
+	 *
+	 * @example
+	 *
+	 * Input: [{id: 1, height: 2}, {id: 2, height: 3}, {id: 3, height: 3}]
+	 * Output: [{id: 2, height: 3}, {id: 3, height: 3}]
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_computeLargestSubsetMaxBy(arrayOfObjects, condition) {
```
Comment (michielmulders) [inline framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js]: This function is difficult to understand if you don't know where it is coming from.
Need to specify what this `condition` parameter is: `peer => peer.height`.  (More or less a lambda / anonymous function)

+ split line 104 into two lines as we use condition as an argument and then call it again on the result.
First, find the peer with the max height and then get the `absoluteMax` out.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+			 * c) Form the peers in b) choose the largest set which has the same blockID.
+			 * 		Ties are broken in favor of smaller blockID.
+			 *
+			 * This set is the set bestPeers and we define B to be the block at the
+			 * tip of the chain of this set of peers.
+			 *
+			 * @link https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#block-synchronization-mechanism
+			 */
+
+			it('should successfully select the best peer according to the steps defined in LIP-0014', () => {
```
Comment (SargeKhan) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: Does it make sense to have a similar structure of CSV file for test scenarios in selecting best peers?

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+	beforeEach(() => {
+		blockSynchronizationMechanism = new BlockSynchronizationMechanism({
+			channel: stubs.channel,
+			modules: stubs.modules,
+		});
+	});
+
+	describe('constructor()', () => {});
+
+	describe('_computeBestPeer()', () => {
```
Comment (shuse2) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: current convetion is
`#_computeBestPeer` I think

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js [hunk 7]
```diff
@@ -0,0 +1,183 @@
+const BlockSynchronizationMechanism = require('../../../../../../../src/modules/chain/synchronizer/block_synchronization_mechanism');
```
Comment (shuse2) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: Requires header

File: framework/test/jest/unit/specs/controller/application_state.spec.js [hunk 8]
```diff
@@ -196,7 +196,7 @@ describe('Application State', () => {
 
 			it('should call get four times', async () => {
 				// Assert
-				expect(spies.get).toHaveBeenCalledTimes(4);
+				expect(spies.get).toHaveBeenCalledTimes(5);
```
Comment (shuse2) [inline framework/test/jest/unit/specs/controller/application_state.spec.js]: why this increase?

File: framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js [hunk 9]
```diff
@@ -1,14 +1,116 @@
...
+	 * From an input list of peers, computes the best peer to continue working with
+	 * according to the set of rules defined in Step 1. of Block Synchroniztion Mechanism
+	 * @link https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#block-synchronization-mechanism
+	 * @param peers
+	 * @return {Array<Object>}
+	 * @private
+	 */
+	_computeBestPeer(peers) {
+		// Largest subset of peers with largest prevotedConfirmedUptoHeight
+		let largestSubset = this._computeLargestSubsetMaxBy(
```
Comment (shuse2) [inline framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js]: prefer no let

File: framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js [hunk 10]
```diff
@@ -1,14 +1,116 @@
...
 	}
 
-	// eslint-disable-next-line class-methods-use-this,no-empty-function
-	async run() {}
+	async run() {
+		this.active = true;
+		const { connectedPeers: peers } = await this.channel.invoke(
+			'network:getNetworkStatus'
+		);
+		if (!peers.length) throw new Error('Connected peers list is empty');
```
Comment (shuse2) [inline framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js]: if syntax should be
```
if () {
}
```

File: framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js [hunk 11]
```diff
@@ -1,14 +1,116 @@
...
+		// eslint-disable-next-line no-unused-vars
+		const bestPeer = this._computeBestPeer(peers);
+		// TODO: handle bestPeer and move on to step 2 defined in
+		// https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#block-synchronization-mechanism
+		// ...
+		this.active = false;
```
Comment (shuse2) [inline framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js]: if it throws above, it will be active forever 😢

File: elements/lisk-p2p/test/unit/peer_selection.ts [hunk 12]
```diff
@@ -25,6 +25,8 @@ describe('peer selector', () => {
 		let peerList = initializePeerInfoList();
 		const nodeInfo: P2PNodeInfo = {
 			height: 545777,
+			lastBlockId: '1234',
```
Comment (shuse2) [inline elements/lisk-p2p/test/unit/peer_selection.ts]: we are not adding any changes to p2p library, but we need to update the test?
I think p2p doesn't know anything about those values.

File: framework/src/modules/chain/chain.js [hunk 13]
```diff
@@ -372,7 +376,15 @@ module.exports = class Chain {
 			rewardMileStones: this.options.constants.REWARDS.MILESTONES,
 			totalAmount: this.options.constants.TOTAL_AMOUNT,
 			blockSlotWindow: this.options.constants.BLOCK_SLOT_WINDOW,
+			synchronizer: new Synchronizer(),
 		});
+
+		this.blocks.synchronizer.addStrategy(
```
[the remark this answers is not in the document]
Comment (limiaspasdaniel) [reply framework/src/modules/chain/chain.js]: IMO Synchronizer shouldn't be aware of the mechanism registered. This means also that if the mechanisms require more dependencies, Synchronizer should need to add them to, depending entirely on its child mechanisms. I think by instantiating, adding dependencies and registering the mechanisms to the Synchronizer at bootstrap time (chain.js) makes more sense in this case. This will prevent modifying the Synchronizer in the  future if mechanisms are altered.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js (hunk 5 above)
Comment (limiaspasdaniel) [reply framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: I think this case is different as the volume of data and input is way less than in BFT case and so it's easier to read as a JSON object. I agree with you that perhaps I can add more tests cases to cover more scenarios.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js (hunk 6 above)
Comment (limiaspasdaniel) [reply framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: Perhaps we can add this to Notion, as I have seen other tests using this format.

File: framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js (hunk 11 above)
Comment (limiaspasdaniel) [reply framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js]: This function will be completed when all the steps in BSM are implemented separately. I will keep this in mind. Thanks

File: elements/lisk-p2p/test/unit/peer_selection.ts (hunk 12 above)
Comment (limiaspasdaniel) [reply elements/lisk-p2p/test/unit/peer_selection.ts]: Agreed, this was before I removed the properties from P2P as @ishantiw told me it is made to support generic ones atm.

**Comment by yatki** [comment]
I think tests were updated after I approved the PR. I'll go through the tests tomorrow morning.

File: framework/test/jest/unit/specs/controller/application_state.spec.js (hunk 8 above)
Comment (limiaspasdaniel) [reply framework/test/jest/unit/specs/controller/application_state.spec.js]: Removed the test case as it is difficult to maintain and will be refactored in #3875

File: framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js (hunk 9 above)
Comment (limiaspasdaniel) [reply framework/src/modules/chain/synchronizer/block_synchronization_mechanism.js]: Done

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js (hunk 7 above)
Comment (limiaspasdaniel) [reply framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: Done

**Comment by limiaspasdaniel** [comment]
@SargeKhan https://github.com/LiskHQ/lisk-sdk/pull/3900#discussion_r299499260
> and also maybe have assertions for all properties. Only checking that height exists doesn't make sense anymore.

We will take care of this in #3875. I don't think adding assertions to individual properties makes sense here. @yatki  already has proposed a way to generalize the behavior of this function so any allowed attribute can be passed as long as it complies with the format.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+	let blockSynchronizationMechanism;
+
+	beforeEach(() => {
+		blockSynchronizationMechanism = new BlockSynchronizationMechanism({
+			channel: stubs.channel,
+			modules: stubs.modules,
+		});
+	});
+
+	describe('#constructor', () => {});
```
Comment (yatki) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: why constructor test suite is empty?

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+
+	describe('#_computeBestPeer', () => {
+		it('should accept an array of peers as input', () => {
+			expect(blockSynchronizationMechanism._computeBestPeer.length).toEqual(1);
```
Comment (yatki) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: prefer to use toHaveLength matcher.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+				prevotedConfirmedUptoHeight: 0,
+			};
+
+			const peers = [
+				{
+					lastBlockId: '12343245',
+					prevotedConfirmedUptoHeight: 1,
+					height: 67,
+					ip: '127.0.0.1',
+				},
```
Comment (yatki) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: I'm just speculating now, maybe putting another peer object into the peers array could improve the test. Because it's returning one of the good peers randomly, right? so, this test doesn't cover that scenario properly. We need to give 2 good and 1 bad peers and make sure that the best peer is always one of those 2 good peers.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+			 * @link https://github.com/LiskHQ/lips/blob/master/proposals/lip-0014.md#block-synchronization-mechanism
+			 */
+
+			it('should successfully select the best peer according to the steps defined in LIP-0014', () => {
+				stubs.modules.blocks.lastBlock = {
+					height: 0,
+					prevotedConfirmedUptoHeight: 0,
+				}; // So ForkChoiceRule.isDifferentChain returns is TRUTHY
+
+				const peers = [
```
Comment (yatki) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: To improve readability, I would move these peers to a fixture file.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js (hunk 14 above)
Comment (limiaspasdaniel) [reply framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: Added the skeleton as these tests will be completed once we have both mechanism in place.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js (hunk 16 above)
Comment (limiaspasdaniel) [reply framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: Good catch. This behavior is being tested at line 85 for a bigger set of peers. so the scenario is covered.

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js (hunk 15 above)
Comment (limiaspasdaniel) [reply framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: Done

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js (hunk 17 above)
Comment (limiaspasdaniel) [reply framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism.spec.js]: Done

File: framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism/peers.js [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+ */
+
+'use strict';
+
+/**
+ * Fixtures used by Block Synchronization Mechanism's Peer selection algorithm
+ * It represents an array of Peers supposedly returned by the network module
+ * @type {*[]}
+ */
+module.exports = [
```
Comment (yatki) [inline framework/test/jest/unit/specs/modules/chain/synchronizer/block_synchronization_mechanism/peers.js]: ❤️ 👍
